### PR TITLE
Fixing cancelling simulations in celery

### LINKED ISSUE #1173: Error on cancelling celery simulation
When number of tasks is greater than number of cores for celery_worker there is error after hitting cancel button. Sometimes cancelling works when number of tasks is greater just by few, but if it's more then it won't work.

### PR DESCRIPTION
**Pull request by jagodek**
Celery_ids are predefined by uuid4 and saved in database before simulation starts. On delete request merge_id and then array of celry_ids are canceled by SIGINT signal.

### CODE REVIEW

**Comment by grzanka** [comment]
@jagodek fix deepsource issues

**Comment by grzanka** [comment]
@jagodek merge recent master changes as well

File: yaptide/persistence/models.py [hunk 1]
```diff
@@ -209,7 +209,9 @@ def update_state(self, update_dict: dict):  # skipcq: PY-R1000
             self.requested_primaries = update_dict["requested_primaries"]
         if "simulated_primaries" in update_dict and self.simulated_primaries != update_dict["simulated_primaries"]:
             self.simulated_primaries = update_dict["simulated_primaries"]
-        if "task_state" in update_dict and self.task_state != update_dict["task_state"]:
+        if "task_state" in update_dict and self.task_state != update_dict["task_state"] and not (
+                self.task_state in [EntityState.COMPLETED.value, EntityState.FAILED.value]
+                and update_dict["task_state"] == EntityState.CANCELED.value):
```
Comment (grzanka) [inline yaptide/persistence/models.py]: the logic here is too complex, readability of this line sucks.
Can you break it down into several nested lines ?
Some comments may be necessary, for example

```suggestion
        if "task_state" in update_dict:
             # check if update is needed
        else:
           # do nothing, as nothing related to task_update came with update dict
            pass
```

File: yaptide/routes/celery_routes.py [hunk 2]
```diff
@@ -156,19 +159,19 @@ def delete(user: UserModel):
...
-
-            return yaptide_response(message="", code=200, content=result)
-
-        return error_internal_response()
+        celery_ids = [
+            task.celery_id for task in tasks
+            if task.task_state in [EntityState.PENDING.value, EntityState.RUNNING.value, EntityState.UNKNOWN.value]
+        ]
+        celery_app.control.revoke(simulation.merge_id, terminate=True, signal="SIGINT")
+        celery_app.control.revoke(celery_ids, terminate=True, signal="SIGINT")
```
Comment (grzanka) [inline yaptide/routes/celery_routes.py:169]:
```suggestion
        celery_app.control.terminate(celery_ids, signal="SIGINT")
```

wouldn't that work as well ?

Comment (grzanka) [inline yaptide/routes/celery_routes.py:169]: write a comment explaining why you terminate first the merge and later the tasks.
I think it makes sense, but needs to be explained

File: yaptide/routes/celery_routes.py [hunk 3]
```diff
@@ -156,19 +159,19 @@ def delete(user: UserModel):
...
+        update_simulation_state(simulation=simulation, update_dict={"job_state": EntityState.CANCELED.value})
+        for task in tasks:
+            if task.task_state in [EntityState.PENDING.value, EntityState.RUNNING.value]:
+                update_task_state(task=task, update_dict={"task_state": EntityState.CANCELED.value})
+
+        terminate_unfinished_tasks.delay(celery_ids=celery_ids, simulation_id=simulation.id)
```
Comment (grzanka) [inline yaptide/routes/celery_routes.py]: is that blocking or non-blocking ?
If non-blocking then why do you later return with "Cancelled successfully" message ?

Comment (grzanka) [reply yaptide/routes/celery_routes.py]: Maybe another state would need to be introduced (in separate issue), like "CANCELLING" which would serve the case of long-running-cancellation state.

File: yaptide/utils/helper_worker.py [hunk 4]
```diff
@@ -1,7 +1,12 @@
...
-                    task_routes={'yaptide.batch.batch_methods*': {
-                        'queue': 'helper'
-                    }})
+                    include=['yaptide.batch.batch_methods', 'yaptide.utils.helper_tasks'],
+                    task_routes={
+                        'yaptide.batch.batch_methods*': {
+                            'queue': 'helper'
+                        },
+                        'yaptide.utils.helper_tasks*': {
+                            'queue': 'helper'
```
Comment (grzanka) [inline yaptide/utils/helper_worker.py]: why putting all of them in the same queue ?

File: yaptide/routes/celery_routes.py (hunk 2 above)
Comment (jagodek) [reply yaptide/routes/celery_routes.py:169]: Revoke is need here because it prevents queued tasks from running in future

Comment (jagodek) [reply yaptide/routes/celery_routes.py:169]: I made intuitive explanation. Not sure if it's right.

File: yaptide/utils/helper_worker.py (hunk 4 above)
Comment (jagodek) [reply yaptide/utils/helper_worker.py]: I made second queue

File: yaptide/routes/celery_routes.py (hunk 3 above)
Comment (jagodek) [reply yaptide/routes/celery_routes.py]: It's non-blocking. I assume that function executes successfully. Testing has proofed that even with large number of tasks and large number of particles it works. I added some timeout there to wait until tasks get canceled after first revoke before trying again.

Comment (jagodek) [reply yaptide/routes/celery_routes.py]: I personally believe that it doesn't matter for frontend user that cancelling will last some time in background after clicking button. But if there is or will be some backend part dependent on it then yes.

File: yaptide/persistence/models.py [hunk 5]
```diff
@@ -163,6 +163,17 @@ def update_state(self, update_dict):
         return db_commit_required
 
 
+def allowed_state_change(current_state, next_state):
```
Comment (grzanka) [inline yaptide/persistence/models.py]: can you add type hinting ?

File: yaptide/persistence/models.py [hunk 6]
```diff
@@ -163,6 +163,17 @@ def update_state(self, update_dict):
...
+    """Ensures that no such change like Completed -> Canceled happens"""
+    return not (current_state in [EntityState.FAILED.value, EntityState.COMPLETED.value]
+                and next_state in [EntityState.CANCELED])
+
+
+def value_changed(current_value, new_value):
```
Comment (grzanka) [inline yaptide/persistence/models.py]: can you add type hinting ? are those values strings ?

File: yaptide/persistence/models.py [hunk 7]
```diff
@@ -205,11 +216,12 @@ def update_state(self, update_dict: dict):  # skipcq: PY-R1000
...
         if self.task_state in (EntityState.COMPLETED.value, EntityState.FAILED.value, EntityState.CANCELED.value):
             return
-        if "requested_primaries" in update_dict and self.requested_primaries != update_dict["requested_primaries"]:
+        if value_changed(self.requested_primaries, update_dict.get("requested_primaries")):
             self.requested_primaries = update_dict["requested_primaries"]
-        if "simulated_primaries" in update_dict and self.simulated_primaries != update_dict["simulated_primaries"]:
+        if value_changed(self.simulated_primaries, update_dict.get("simulated_primaries")):
             self.simulated_primaries = update_dict["simulated_primaries"]
-        if "task_state" in update_dict and self.task_state != update_dict["task_state"]:
+        if value_changed(self.task_state, update_dict.get("task_state")) and allowed_state_change(
```
Comment (grzanka) [inline yaptide/persistence/models.py:223]: very nice !!!

File: yaptide/routes/celery_routes.py (hunk 2 above)
Comment (grzanka) [reply yaptide/routes/celery_routes.py:169]: Thats correct

File: yaptide/utils/helper_tasks.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+from time import sleep
+from yaptide.utils.helper_worker import celery_app
+
+
+@celery_app.task
+def terminate_unfinished_tasks(simulation_id):
```
Comment (grzanka) [inline yaptide/utils/helper_tasks.py]: can you add type hints ?

File: yaptide/utils/helper_tasks.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    while number_of_tasks != previous_number_of_tasks:
+        previous_number_of_tasks = number_of_tasks
+        sleep(1)
+        number_of_tasks = get_tasks_from_celery(simulation_id)
+    celery_app.control.revoke([celery_pair['celery_id'] for celery_pair in get_tasks_from_celery(simulation_id)],
+                              terminate=True,
+                              signal="SIGINT")
+
+
+def get_tasks_from_celery(simulation_id):
```
Comment (grzanka) [inline yaptide/utils/helper_tasks.py]: can you add type hints ?

**Comment by grzanka** [comment]
https://github.com/yaptide/yaptide/pull/1140 was merged just now, time to sync this branch with recent master

**Review by grzanka — APPROVED** [review]

**Pull request merged**